vm/chef/cookbooks/mariadb/files/mariadb-utils

```diff
@@ -0,0 +1,323 @@
+    scp -o StrictHostKeyChecking=no "${backup_filename}" \
+      "${replication_user}@${replica}:${backup_folder}"
+    echo "Backup sent to ${replica}."
+  done
+}
+
+#######################################
+# Setup a master MariaDB node.
+# Arguments:
+#   Replicas list separated by spaces.
+#   root user password.
+#   replication user password.
+#   Enable root access from any host or not.
+#######################################
+function setup_master() {
+  local -r replicas="$1"
+  local -r root_password="$2"
+  local -r replication_password="$3"
+  local -r enable_root_anywhere="$4"
+
+  apply_master_replication_config
+
+  # Start and await service to be up
+  systemctl start mariadb
+  until mysqladmin ping; do sleep 1; done
+
+  create_replication_users "${replicas}" "${replication_password}"
+  set_root_user "${root_password}" "${enable_root_anywhere}"
+
+  create_backup
+  await_for_replicas_ssh "${replicas}"
+  send_backup_to_replicas "${replicas}"
+
+  run_query "UNLOCK TABLES;"
+}
+
+#######################################
+# Applies the replica configuration to a local MariaDB setup.
+#######################################
+function apply_replica_replication_config() {
+  local -r config_file="/etc/mysql/conf.d/serverid.cnf"
+
+  printf '%s\n' \
+    "[mysqld]"   \
+    "server-id=${SERVER_ID}"   \
+    "replicate-do-db=${DATABASE_NAME}" > "${config_file}"
+}
+
+#######################################
+# Await if MariaDB service is up at specific node.
+# Arguments:
+#  Node hostname or IP.
+#######################################
+function await_for_node() {
+  local node="$1"
+
+  emit_msg "Checking if master node is up..."
+  timeout --preserve-status 300 bash -c "until echo > /dev/tcp/${node}/3306; do sleep 2; done"
+  if [[ "$?" -ne 0 ]]; then
+      exit 1
+  fi
+}
+
+#######################################
+# Enable SSH access to replication user without password
+# and only from internal master node IP.
+#######################################
+function enable_ssh_on_replica() {
+    # Add replication user and tmp folder
+    adduser --disabled-password --gecos "" -u 2001 ${replication_user} \
+      && create_tmp_folder
+
+    # Remove password
+    sed -i -re "s/^${replication_user}:[^:]+:/${replication_user}::/" \
+      /etc/passwd /etc/shadow
+
+    # Configure pam.d
+    sed -i '/@include common-auth/r patch-pam-ssh' /etc/pam.d/sshd
+    sed -i '/@include common-auth/d' /etc/pam.d/sshd
+
+    # Configure sshd_config
+    sed -i-re 's/^PasswordAuthentication no$/# PasswordAuthentication no/g' \
+        /etc/ssh/sshd_config
+    patch="$(cat patch-ssh | envsubst "\$MASTER_NODE_IP")"
+    echo "${patch}" >> /etc/ssh/sshd_config
+
+    systemctl restart ssh
+}
+
+#######################################
+# Await until backup file is available locally.
+#######################################
+function await_for_backup_file() {
+  local file_count=0
+
+  while [[ "${file_count}" -eq 0 ]]; do
```

Comment:
maybe check for specific name? what if there are already some files from previous backup?

Reply:
Done.